agent/util-scripts/pbench-tool-meister-start

```diff
@@ -0,0 +1,542 @@
+import redis
+import logging
+from distutils.spawn import find_executable
+from pathlib import Path
+
+
+# Port number is "One Tool" in hex 0x17001
+# FIXME: move to common area
+redis_port = 17001
+
+# Redis server configuration template for pbench's use
+redis_conf_tmpl = """bind {hostnames}
+daemonize yes
+dir {tm_dir}
+dbfilename pbench-redis.rdb
+logfile {tm_dir}/redis.log
+loglevel notice
+pidfile {tm_dir}/redis_{redis_port:d}.pid
+port {redis_port:d}
+"""
+
+# FIXME: this should be moved to a shared area
+channel = "tool-meister-chan"
+
+# Maximum time to wait for the Redis server to respond.
+REDIS_MAX_WAIT = 60
+
+
+class ToolGroup(object):
+    tg_prefix = "tools-v1"
+
+    def __init__(self, group):
+        self.group = group
+        _pbench_run = os.environ["pbench_run"]
+        self.tg_dir = Path(_pbench_run, f"{self.tg_prefix}-{self.group}").resolve(
+            strict=True
+        )
+        if not self.tg_dir.is_dir():
+            raise Exception(
+                f"bad tool group, {group}: directory {self.tg_dir} does not exist"
+            )
+
+        # __trigger__
+        try:
+            with (self.tg_dir / "__trigger__").open("r") as fp:
+                _trigger = fp.read()
+        except OSError as ex:
+            if ex.errno != errno.ENOENT:
+                raise
+            # Ignore missing trigger file
+            self.trigger = None
+        else:
+            if len(_trigger) == 0:
+                # Ignore empty trigger file contents
+                self.trigger = None
+            else:
+                self.trigger = _trigger
+
+        # toolnames - Dict with tool name as the key, dictionary with host
+        # names and parameters for each host
+        self.toolnames = {}
+        # hostnames - Dict with host name as the key, dictionary with tool
+        # names and parameters for each tool
+        self.hostnames = {}
+        self.labels = {}
+        for hdirent in os.listdir(self.tg_dir):
+            if hdirent == "__trigger__":
+                # Ignore handled above
+                continue
+            if not (self.tg_dir / hdirent).is_dir():
+                # Ignore wayward non-directory files
+                continue
+            # We assume this directory is a hostname.
+            host = hdirent
+            if host not in self.hostnames:
+                self.hostnames[host] = {}
+            for tdirent in os.listdir(self.tg_dir / host):
+                if tdirent == "__label__":
+                    with (self.tg_dir / host / tdirent).open("r") as fp:
+                        self.labels[host] = fp.read()
+                    continue
+                if tdirent.endswith("__noinstall__"):
+                    # FIXME: ignore "noinstall" for now, tools are going to be
+                    # in containers so this does not make sense going forward.
+                    continue
+                tool = tdirent
+                with (self.tg_dir / host / tool).open("r") as fp:
```

Comment:
(self.tg_dir / host / tool).read_text() is easier to use here (part of pathlib)

Reply:
Fixed.